formats/opentype.fathom

```diff
@@ -401,6 +413,690 @@ let device_or_variation_index_table = overlap {
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: Chained Sequence Context Format 3](https://docs.microsoft.com/en-us/typography/opentype/spec/chapter2#chained-sequence-context-format-3-coverage-based-glyph-contexts)
+let chained_sequence_context_format_3 = {
+    /// The start of the table
+    table_start <- stream_pos,
+    /// Number of glyphs in the backtrack sequence
+    backtrack_glyph_count <- u16be,
+    /// Array of offsets to coverage tables for the backtrack sequence
+    backtrack_coverages <- array16 backtrack_glyph_count (offset16 table_start coverage_table),
+    /// Number of glyphs in the input sequence
+    input_glyph_count <- u16be,
+    /// Array of offsets to coverage tables for the input sequence
+    input_coverage_tables <- array16 input_glyph_count (offset16 table_start coverage_table),
+    /// Number of glyphs in the lookahead sequence
+    lookahead_glyph_count <- u16be,
+    /// Array of offsets to coverage tables for the lookahead sequence
+    lookahead_coverages <- array16 lookahead_glyph_count (offset16 table_start coverage_table),
+    /// Number of SequenceLookupRecords
+    seq_lookup_count <- u16be,
+    /// Array of SequenceLookupRecords
+    seq_lookup_records <- array16 seq_lookup_count sequence_lookup_record,
+};
+
+let chained_sequence_context = {
+    /// Format identifier
+    format <- u16be,
+    /// Format specific substitutions
+    subst <- match format {
+        1 => sequence_context_format1,
+        2 => sequence_context_format2,
+        3 => sequence_context_format3,
+        _ => unknown_table,
+    }
+};
+
+// # Lookup sub-tables
+
+/// # LookupType 1: Single Substitution Subtable
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: LookupType 1](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#lookuptype-1-single-substitution-subtable)
+let single_substitution = {
+    /// The start of the sub-table
+    table_start <- stream_pos,
+    /// Format identifier
+    subst_format <- u16be,
+    /// Coverage table
+    coverage <- offset16 table_start coverage_table,
+
+    subst <- match subst_format {
+        1 => {
+            /// Add to original glyph ID to get substitute glyph ID
+            delta_glyph_id <- s16be,
+        },
+        2 => {
+            /// Number of glyph IDs in the substitute_glyph_ids array
+            glyph_count <- u16be,
+            /// Array of substitute glyph IDs — ordered by Coverage index
+            substitute_glyph_ids <- array16 glyph_count u16be,
+        },
+        _ => unknown_table
+    },
+};
+
+/// LookupType 2: Multiple Substitution Subtable
+///
+/// ## References
+///
+/// - [Microsoft's OpenType Spec: LookupType 2](https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#lookuptype-2-multiple-substitution-subtable)
+let multiple_substitution = (
+    let sequence_table = {
+        /// Number of glyph IDs in the substitute_glyph_ids array. This must always be greater than
+        /// 0.
```

Comment:
Does the spec still say multiple subst can't be zero? Because fonts and implementations agree it can, and I thought they revised that for the spec beta.

Reply:
That text is copied from the spec. Perhaps the beta edit hasn't been released yet.

Reply:
Seems that there were patches to a bunch of tools but the actual issue on the spec remains open: 

https://github.com/MicrosoftDocs/typography-issues/issues/673

It appears to be with Microsoft (since 2020):

https://github.com/MicrosoftDocs/typography-issues/issues/673#issuecomment-718114268